### PR TITLE
Gift Entry: template level customizable batch table header

### PR DESCRIPTION
**Pull request by psadie-sfdo**
Adds the ability for users to select their template level default batch table headers when using Gift Entry in Batch mode.
- Custom Labels
-- geBodyBatchTableColumnsPart1
-- geBodyBatchTableColumnsPart2
-- geTabBatchSettings
-- geTabBatchTableColumns
-- geErrorFLSBatchTableColumns
# Critical Changes

# Changes
## Gift Entry Updates
- We updated Template Builder so users can choose which fields to include the batch gift table.
# Issues Closed

# Community Ideas Delivered

# New Metadata

- New Fields
  - DataImportBatch__c.Batch_Table_Columns_JSON__c
# Deleted Metadata

### CODE REVIEW

**Comment by psadie-sfdo** [comment]
**lurch: attach W-041565

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000007haZ0QAI" target="_blank">W-041565</a>

File: src/labels/CustomLabels.labels [hunk 1]
```diff
@@ -5487,6 +5487,14 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
         <shortDescription>Body below the Advanced Form Fields header</shortDescription>
         <value>For more advanced gift entry options, select fields from the following Object Groups.</value>
     </labels>
+    <labels>
+        <fullName>geBodyBatchTableColumns</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Body below the Advanced Form Fields header</shortDescription>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5495]:
```suggestion
        <shortDescription>Description for Batch Table Columns, part 1</shortDescription>
```

**Comment by ptewson-sfdo** [comment]
@psadie-sfdo One thing we will have to address is how to handle batches that were set up prior to introducing the batch table customizations - I just deployed this branch, and when opening my pre-existing batch I was greeted with a "null input to JSON parser" toast and "[NoErrorObjectAvailable] Script error" popup as well

File: src/labels/CustomLabels.labels (hunk 1 above)
Comment (psadie-sfdo) [reply src/labels/CustomLabels.labels]: fixed!

File: src/labels/CustomLabels.labels [hunk 2]
```diff
@@ -5903,7 +5929,7 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
         <language>en_US</language>
         <protected>true</protected>
         <shortDescription>Label for column in batch gift entry table that links to the matched Gift.</shortDescription>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5931]:
```suggestion
        <shortDescription>Donation name (Opportunity or Payment) in Gift Entry batch table</shortDescription>
```
```suggestion
        <shortDescription>Label for batch gift table column that links to gift (Opportunity or Payment)</shortDescription>
```

Comment (melissabarber) [inline src/labels/CustomLabels.labels:5931]:
```suggestion
        <shortDescription>Label for batch gift table column that links to gift (Opportunity or Payment)</shortDescription>
```

File: src/labels/CustomLabels.labels [hunk 3]
```diff
@@ -5927,7 +5953,7 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
         <language>en_US</language>
         <protected>true</protected>
         <shortDescription>Label for column in batch gift entry table that links to the Account or Contact.</shortDescription>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5955]:
```suggestion
        <shortDescription>Label for batch gift table column that links to the Account or Contact.</shortDescription>
```

**Comment by melissabarber** [comment]
@psadie-sfdo @ptewson-sfdo added a few minor shortDescription change suggestions, but other than that, the PR looks good from docs point of view. Though, there's been enough back and forth about this work, I wouldn't mind seeing a demo before this PR is merged. Probably loop Grace in on that as well. Thanks!

File: src/classes/BGE_DataImportBatchEntry_CTRL.cls [hunk 4]
```diff
@@ -394,6 +396,22 @@ public with sharing class BGE_DataImportBatchEntry_CTRL {
         return [SELECT Count() FROM DataImport__c WHERE NPSP_Data_Import_Batch__c =: recordId];
     }
 
+    /*******************************************************************************************************
+    * @description Gets the Gift Entry defined list of Batch Table Columns as JSON.
+    * @param recordId the ID of the NPSP_Data_Import_Batch__c
+    * @return String: List of Batch Table Columns as JSON
+    */
+    private static String[] getGiftEntryDefinedBatchTableColumns(Id recordId) {
```
Comment (kenlewis-sfdo) [inline src/classes/BGE_DataImportBatchEntry_CTRL.cls:404]: The JS file of the GeGiftEntryFormApp [uses getRecord](https://github.com/SalesforceFoundation/NPSP/blob/46d9296bfb752e19fcf9fda42e657dc522190047/src/lwc/geGiftEntryFormApp/geGiftEntryFormApp.js#L578) to import some fields on the batch already.  Could the App use that to retrieve the value of the new Batch Table Columns field and pass it down to the table component?  Then this method wouldn't be needed and the DataImportModel wouldn't need to be altered.

File: src/lwc/utilTemplateBuilder/utilTemplateBuilder.js [hunk 5]
```diff
@@ -130,6 +132,7 @@ const EXCLUDED_BATCH_HEADER_FIELDS = [
     DI_BATCH_DEFAULTS_INFO.fieldApiName,
     DI_BATCH_GIFT_ENTRY_VERSION_INFO.fieldApiName,
     DI_BATCH_FORM_TEMPLATE_INFO.fieldApiName,
+    DI_BATCH_TABLE_COLUMNS_FIELD.fieldApiName,
```
Comment (ptewson-sfdo) [inline src/lwc/utilTemplateBuilder/utilTemplateBuilder.js:135]: Could we use this opportunity to include Created Date, Created By ID, Last Modified Date, and Last Modified by ID in the excluded fields here? These are generally restricted due to being read-only but some users see them appear when the have enabled the ability to edit system audit fields on their profile

**Comment by melissabarber** [comment]
@psadie-sfdo will get sign-off from Grace, but docs-wise, this PR is good to go. Labels, help text, and description look good. FYI @ptewson-sfdo & @vmartinov-sforg.

File: src/lwc/utilTemplateBuilder/utilTemplateBuilder.js (hunk 5 above)
Comment (psadie-sfdo) [reply src/lwc/utilTemplateBuilder/utilTemplateBuilder.js:143]: Added a collection of system fields below!

**Comment by psadie-sfdo** [comment]
> Sorry for the delay; it took a little while longer than anticipated to spin up a scratch org. Two questions for you @psadie-sfdo:
> 
> * Should the visible fields include the four recommended ones (Donation Name, Donor Name, Status, Failure Information)?
> * The alignment of "Available Fields" and "Visible Fields" looks a little tight against the selection boxes and a smidge too far to the left. Did you run it by Grace?
> 
> Gift table column selection
> 
> CC: @vmartinov-sforg

- We've opted not to make changes to the default template that we ship with in this case. But happy to revisit that decision if we feel strongly that we should. Alternatively we could add additional logic to force those selection of those fields on whenever we don't find any selected headers when editing an existing template. Thoughts @vmartinov-sforg @kenlewis-sfdo?
- Grace has seen this, but we didn't take a deep look at the position and paddings of those labels. My initial thought is that this might actually be a bit challenging to adjust as we're using a base lightning component here. I'll check in with Grace to get her thoughts regardless though!

File: src/labels/CustomLabels.labels [hunk 6]
```diff
@@ -5512,6 +5512,32 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
...
+    <labels>
+        <fullName>geBodyBatchTableColumnsPart2</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Description for Batch Table Columns, part 2</shortDescription>
+        <!-- "Donor Name" and "Donation Name" texts in this label have related custom labels. -->
+        <!-- Related custom labels are geDonorColumnLabel annd geDonationColumnLabel -->
+        <value>
+        The recommended fields are included.
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5532]:
```suggestion
        We recommend including the following fields:
```

File: src/labels/CustomLabels.labels [hunk 7]
```diff
@@ -5512,6 +5512,32 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
...
+        &lt;/br&gt;
+        &lt;ul&gt;
+            &lt;li&gt; Donor Name links to the Organization Account or Contact giving the donation. &lt;/li&gt;
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5535]:
```suggestion
            &lt;li&gt; Donor Name—links to the Organization Account or Contact giving the donation. &lt;/li&gt;
```

File: src/labels/CustomLabels.labels [hunk 8]
```diff
@@ -5512,6 +5512,32 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
...
+            &lt;li&gt; Donation Name links to the Opportunity or Payment for the donation. &lt;/li&gt;
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5536]:
```suggestion
            &lt;li&gt; Donation Name—links to the Opportunity or Payment for the donation. &lt;/li&gt;
```

File: src/labels/CustomLabels.labels [hunk 9]
```diff
@@ -5512,6 +5512,32 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
...
+            &lt;li&gt; Status and Failure Information give you information about batch processing. &lt;/li&gt;
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5537]:
```suggestion
            &lt;li&gt; Status and Failure Information—give you information about batch processing. &lt;/li&gt;
```

**Pull request merged**